### PR TITLE
expression: fix date cmp date

### LINKED ISSUE #4069: TiDB does not use compound primary key while scanning table
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

Create table syntax:

```
CREATE TABLE `item` (
  `snapshot_date` date NOT NULL,
  `itemid` int(11) NOT NULL,
  `catid` int(11) DEFAULT NULL,
  `sub_catid` int(11) DEFAULT NULL,
  `third_catid` int(11) DEFAULT NULL,
  `shopid` int(11) DEFAULT NULL,
  `brand` text DEFAULT NULL,
  `name` text NOT NULL,
  `price_max` float DEFAULT NULL,
  `price_min` float DEFAULT NULL,
  `sold` int(11) DEFAULT NULL,
  `stock` int(11) DEFAULT NULL,
  `rating_count` int(11) DEFAULT NULL,
  `data` longtext NOT NULL,
  PRIMARY KEY (`snapshot_date`,`itemid`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8_bin;
```

Query below takes very long time and sometimes timed out:

```
select itemid, catid from item where snapshot_date='2017-08-02' and itemid > 293767706 order by itemid limit 10
```

Here is the result of `explain`:

```
+----------------+--------------+----------------+------+--------------------------------------------------------------------+-------------------+
| id             | parents      | children       | task | operator info                                                      | count             |
+----------------+--------------+----------------+------+--------------------------------------------------------------------+-------------------+
| TableScan_8    | Selection_9  |                | cop  | table:item, range:(-inf,+inf), keep order:false                    |        43241219.2 |
| Selection_9    |              | TableScan_8    | cop  | gt(shopee_id.item.itemid, 293767706)                               |        43241219.2 |
| TableReader_10 | Selection_2  |                | root | data:Selection_9                                                   |        43241219.2 |
| Selection_2    | TopN_6       | TableReader_10 | root | eq(cast(shopee_id.item.snapshot_date), 2017-08-02 00:00:00.000000) | 34592975.36000001 |
| TopN_6         | Projection_3 | Selection_2    | root |                                                                    |                10 |
| Projection_3   |              | TopN_6         | root | shopee_id.item.itemid, shopee_id.item.catid                        |                10 |
+----------------+--------------+----------------+------+--------------------------------------------------------------------+-------------------+
```

2. What did you expect to see?

The query should finish within no more than 2 seconds. 

3. What did you see instead?

The query takes very long time and sometimes timed out.

4. What version of TiDB are you using (`tidb-server -V`)?

The latest RC4.

```
Git Commit Hash: bae8fe46f9305cdfff52cf18d5701ae7ead14d75
UTC Build Time:  2017-08-06 01:02:35
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL @hanfei1991 @zz-jason @shenli 

This should fix #4069

### CODE REVIEW

**Comment by zz-jason** [comment]
should we add `baseDateBuiltinfunc`, `baseDatetimeBuiltinFunc` ... and corresponding `evalDate`, `evalDatetime` ... ?

**Comment by shenli** [comment]
CI failed.

**Comment by XuHuaiyu** [comment]
@zz-jason we can take it into consideration, but it's not really necessary for this bug fix.

**Comment by XuHuaiyu** [comment]
PTAL @shenli @zz-jason

File: expression/builtin_op.go [hunk 1]
```diff
@@ -732,7 +732,7 @@ func (c *isNullFunctionClass) getFunction(args []Expression, ctx context.Context
 		sig = &builtinDecimalIsNullSig{baseIntBuiltinFunc{bf}}
 	case tpReal:
 		sig = &builtinRealIsNullSig{baseIntBuiltinFunc{bf}}
-	case tpTime:
+	case tpDatetime, tpDate, tpTimestamp:
```
Comment (zz-jason) [inline expression/builtin_op.go]: `argTp` can never be `tpDate`, `tpTimestamp`, `case tpDatetime` is enough

File: plan/typeinfer_test.go [hunk 2]
```diff
@@ -932,9 +932,9 @@ func (s *testPlanSuite) createTestCase4Miscellaneous() []typeInferTestCase {
 		{"any_value(c_float_d)", mysql.TypeFloat, charset.CharsetBin, mysql.BinaryFlag, 12, types.UnspecifiedLength},
 		{"any_value(c_double_d)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, 22, types.UnspecifiedLength},
 		{"any_value(c_decimal)", mysql.TypeNewDecimal, charset.CharsetBin, mysql.BinaryFlag, 6, 3},
-		{"any_value(c_datetime)", mysql.TypeDatetime, charset.CharsetBin, mysql.BinaryFlag, 22, 2},
+		{"any_value(c_datetime)", mysql.TypeDatetime, charset.CharsetUTF8, mysql.NotNullFlag, 22, 2},
```
Comment (zz-jason) [inline plan/typeinfer_test.go]: in mysql:
```sql
MySQL > select any_value(c_datetime) from t;
Field   1:  `any_value(c_datetime)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  utf8_general_ci (33)
Length:     66
Max_length: 0
Decimals:   2
Flags:
```
result type do not have `NOT NULl` flag

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Comment by XuHuaiyu** [comment]
ptal @shenli @hanfei1991

**Comment by hanfei1991** [comment]
Can we merge tpDate and tpDatetime ?

**Comment by XuHuaiyu** [comment]
I'm not really sure, I'll try it.

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Pull request merged**